File: MATLAB/stone_id/+ops/@import/import.m
classdef import
    %IMPORT Import datastore
    %
    % Methods:
    %   ops.import.read - Import OSTA data
    %
    %

    methods (Static)

        function [value, summary] = read( dslocation, options )

            arguments
                dslocation (1,1) string = ""
                options.Class (1,1) string {mustBeMember(options.Class, ["All", "ABC", "D"])} = "All"
            end

            if ~isfolder( dslocation ) && ~isfile( dslocation )
                error("Location of datastore not found.")
            end

            files = matlab.io.datastore.DsFileSet( dslocation, ...
                "FileExtensions", ".xlsx").resolve.FileName;

            switch options.Class
                case "ABC"
                    tF = contains(files, "all_data");
                    files = files(tF);
                case "D"
                    tF = contains(files, "Class_D");
                    files = files(tF);
                otherwise
                    % continue
            end

            ds = spreadsheetDatastore( files, ...
                "FileExtensions", ".xlsx", ...
                "IncludeSubfolders", true, ...
                "TextType", "string", ...
                "ReadVariableNames", true, ...
                "VariableNamingRule", "preserve");

            data = readall(ds);

            data.Feature = [];
            tF = ismember(data.Properties.VariableNames, ["Sample ID", "Response"]);
            data.Properties.VariableNames(~tF) = "Feature " + data.Properties.VariableNames(~tF);

            data.Response = categorical(data.Response);

            value = data;

            summary = groupsummary( value, "Response" );

        end

        function [value, summary] = readCSV( dsLocation, options )
            % Input validation
            arguments
                dsLocation (1,:) string = "";
                options.FileIndex (1,1) double = inf;
                options.SpectrometerType (1,1) ...
                    {mustBeMember(options.SpectrometerType, ["XL", "CL"])} = ...
                    "XL";
                options.IncludeSubfolders (1,1) logical = false;
            end

            % Get all files present in location
            files = matlab.io.datastore.DsFileSet( fullfile(dsLocation), ...
                "FileExtensions", ".csv", ...
                "IncludeSubfolders", options.IncludeSubfolders).resolve.FileName;

            if isempty(files)
                disp("No files to read.")
                value = [];
                summary = [];
                return;
            end

            if strcmp(options.SpectrometerType, "CL")
                files = files(contains(files, "CL"));
            elseif strcmp(options.SpectrometerType, "XL")
                files = files(contains(files, "XL"));
            end

            if isempty(files)
                fprintf("No files to read. You have selected %s " + ...
                    "spectrometer type. Try changing it to %s type.", ...
                    options.SpectrometerType, ...
                    setdiff(["XL", "CL"], options.SpectrometerType));
                value = [];
                summary = [];
                return;
            end

            % What files to read?
            if options.FileIndex==inf
                fileIndex = 1:length(files);
            else
                fileIndex = options.FileIndex;
            end

            % Initialize data table
            % Meta data columns
            value = cell(1, length(fileIndex));
            % Loop through every file
            for currFileIdx = fileIndex(:)'
                try
                    fn = files(currFileIdx);

                    %% Read in spectroscopy data
                    data = readtable( fn , ...
                        'Range', 'A60');
                    data(:,1) =  [] % remove wavelength column
                    data = rows2vars(data);


                    data(:,1) = [];
                    RotationCount = data{:,1};
                    data(:,1) = [];

                  %  data(:,1) = [];
                    responseRegression = data{:,1};
                    data(:,1) = [];

                    % consider only first 2048 readings for uniformity
                    data = data(:,1:2048);

                    data.Properties.VariableNames = ...
                        "Feature " + string(1:width(data));

                    % assign response data
                    data.ResponseRegression = round(responseRegression);
                    data.Rotation = round(RotationCount);

                    %% Assign metadata
                    opts = delimitedTextImportOptions('NumVariables', 2, ...
                        'DataLines', [1 58], ...
                        'VariableNamesLine', 0, ...
                        'VariableTypes', ["string", "string"], ...
                        'ExtraColumnsRule', 'ignore' );
                    metaData = readtable(fn, opts);
                    metaData = rows2vars(metaData, ...
                        'VariableNamingRule', 'preserve', ...
                        'VariableNamesSource', 1);

                    % Extract light source
                    lightSource = metaData.("Light Source"){1};
                    data.LightSource = repmat(string(lightSource), height(data), 1);

                    % Extract fiber type
                    probeSize = metaData.("Fiber Type"){1};
                    data.ProbeSize = repmat(string(probeSize), height(data), 1);

                    % Extract aiming beam status
                    aimingBeam = metaData.("Aiming Beam Status"){1};
                    data.AimingBeam =repmat(string(aimingBeam), height(data), 1);

                    % Extract stone info
                    targetType = metaData.("Target Type"){1};
                    data.TargetType = repmat(string(targetType), height(data), 1);

                    % Extract stone info
                    targetNumber = metaData.("Target Number"){1};
                    data.TargetNumber = repmat(string(targetNumber), height(data), 1);

                    % Rotation = metaData.("Rotation Count of Target"){1};
                    % data.Rotation = repmat(string(Rotation), height(data), 1);

                    % Extract integration time used
%                     integrationTimeUsed = metaData.("Integration Time Used (mS)"){1};
                    [~,name,~] = fileparts(fn);
                    integrationTimeUsed = extractBefore(name, "_");
                    data.IntegrationTimeUsed = repmat(integrationTimeUsed, height(data), 1);

                    % Extract spectrometer used
                    spectrometerUsed = metaData.("Spec Used");
                    data.SpectrometerUsed = repmat(string(spectrometerUsed), height(data), 1);

                    % Extract filename
                    data.FileName = repmat(fn, height(data), 1);

                    value{ismember(fileIndex, currFileIdx)} = data;
                catch
                    fprintf('Unable to read from file: %s', fn);
                    continue;
                end
            end

            value = vertcat(value{:});
            metaDataVars = value.Properties.VariableNames(...
                ~contains(value.Properties.VariableNames, "Feature"));
            summary = groupsummary( value, metaDataVars );
        end


    end

end

File: Python/TS_MatlabMigration/ops/import.py
import os
import pandas as pd
import numpy as np
import glob


class Import:

    @staticmethod
    def read(dslocation="", class_option="All"):
        # Validate inputs
        if not os.path.isdir(dslocation) and not os.path.isfile(dslocation):
            raise FileNotFoundError("Location of datastore not found.")

        # Get files with .xlsx extension
        files = glob.glob(os.path.join(dslocation, '*.xlsx'))

        if class_option == "ABC":
            files = [f for f in files if "all_data" in f]
        elif class_option == "D":
            files = [f for f in files if "Class_D" in f]

        # Create a pandas dataframe for storing data
        data_frames = []

        for file in files:
            ds = pd.read_excel(file)
            data_frames.append(ds)

        data = pd.concat(data_frames, ignore_index=True)

        # Process the data
        data = data.rename(columns=lambda x: "Feature " + x if x not in ["Sample ID", "Response"] else x)
        data["Response"] = pd.Categorical(data["Response"])

        # Generate summary
        summary = data.groupby("Response").size().reset_index(name="count")

        return data, summary

    @staticmethod
    def readCSV(ds_location="", file_index=float('inf'), spectrometer_type="XL", include_subfolders=True):
        # Get all files with .csv extension
        if include_subfolders:
            files = glob.glob(os.path.join(ds_location, '**/*.csv'), recursive=True)
        else:
            files = glob.glob(os.path.join(ds_location, '*.csv'))

        if not files:
            print("No files to read.")
            return [], []

        # Filter files based on spectrometer type
        if spectrometer_type == "CL":
            files = [f for f in files if "CL" in f]
        elif spectrometer_type == "XL":
            files = [f for f in files if "XL" in f]

        if not files:
            print(f"No files to read. You selected {spectrometer_type} spectrometer type.")
            return [], []

        # Select files to read based on file index
        if file_index == float('inf'):
            file_indices = range(len(files))
        else:
            file_indices = [file_index]

        all_data = []

        for idx in file_indices:
            try:
                file_path = files[idx]
                # Read spectroscopy data, skipping first rows (like A64 in MATLAB)
                data = pd.read_csv(file_path, skiprows=63)
                data = data.iloc[:, 1:].T  # remove wavelength column and transpose

                # Use only first 2048 readings for uniformity
                data = data.iloc[:, :2048]

                # Rename columns as "Feature 1", "Feature 2", ..., "Feature 2048"
                data.columns = [f"Feature {i + 1}" for i in range(data.shape[1])]

                # Metadata extraction (adjust this based on your metadata structure)
                meta_data = pd.read_csv(file_path, nrows=58)
                light_source = meta_data.loc[0, "Light Source Type"]
                probe_size = meta_data.loc[0, "Fiber Type"]
                aiming_beam = meta_data.loc[0, "Aiming Beam Status"]
                target_type = meta_data.loc[0, "Target Type"]
                target_number = meta_data.loc[0, "Target Number"]
                integration_time_used = meta_data.loc[0, "Integration Time Used (mS)"]
                spectrometer_used = meta_data.loc[0, "Spec Used"]

                # Add metadata columns to the data
                data["LightSource"] = light_source
                data["ProbeSize"] = probe_size
                data["AimingBeam"] = aiming_beam
                data["TargetType"] = target_type
                data["TargetNumber"] = target_number
                data["IntegrationTimeUsed"] = integration_time_used
                data["SpectrometerUsed"] = spectrometer_used
                data["FileName"] = os.path.basename(file_path)

                all_data.append(data)

            except Exception as e:
                print(f"Unable to read file: {file_path}. Error: {e}")
                continue

        final_data = pd.concat(all_data, ignore_index=True)

        # Metadata columns used for grouping
        meta_data_vars = [col for col in final_data.columns if not col.startswith("Feature")]
        summary = final_data.groupby(meta_data_vars).size().reset_index(name="count")

        return final_data, summary
